### PR TITLE
Source server variables from IIS when running ANCM in-proc

### PR DESCRIPTION
**Pull request by mderriey**
- Source server variables from IIS when running ANCM in-proc

Addresses #6013

### CODE REVIEW

**Comment by mderriey** [comment]
This is the first version to get something out there before asking too many questions.

Right now it doesn't honor the `UriMatchPart` parameter and returns IIS variable as-is.

Let me know if we'd like to support that or if you see anything else.

File: src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.csproj
Comment (anurse) [inline src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.csproj]: Hrm, it's unfortunate that we need to reference `.Server.IIS` just to get `IServerVariablesFeature`. Could we consider relocating that feature interface? It's very simple.

File: src/Middleware/Rewrite/test/IISUrlRewrite/ServerVariableTests.cs
Comment (anurse) [inline src/Middleware/Rewrite/test/IISUrlRewrite/ServerVariableTests.cs]: I'd prefer just creating a concrete `TestServerVariablesFeature` class that takes a dictionary and exposes it via the accessor. It's trivial and means we don't need more `Mock`s. In general, I lean towards using actual types rather than more dynamic code gen :)

```csharp
public class TestServerVariablesFeature: IServerVariablesFeature
{
    private readonly Dictionary<string, string> _variables;

    public TestServerVariablesFeature(IDictionary<string, string> variables)
    {
        _variables = variables;
    }

    public string this[string variableName]
    {
        get => _variables[variableName];
        set => _variables[variableName] = value;
    }
}
```

**Review by analogrelay — COMMENTED** [review]
Looks OK to me. Minor suggestion around using a real type instead of a mock. Also not sure about adding the dependency. Now that it's one big happy `<FrameworkReference>`, it's probably not a big deal, but having to depend on `Servers.IIS` from `Rewrite` feels a little off.

**Review by analogrelay — COMMENTED** [review]
Looks OK to me. Minor suggestion around using a real type instead of a mock. Also not sure about adding the dependency. Now that it's one big happy `<FrameworkReference>`, it's probably not a big deal, but having to depend on `Servers.IIS` from `Rewrite` feels a little off.

File: src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.csproj
Comment (jkotalik) [reply src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.csproj]: I can go either way on this change. In one hand, the ServerVariable feature is pretty much purely an IIS feature, so it makes sense to live in that namespace. On the other hand, the feature is simple. 

I don't think it's a huge deal having Rewrite depend on IIS.

File: src/Middleware/Rewrite/src/Internal/PatternSegments/IISServerVariableSegment.cs
Comment (jkotalik) [inline src/Middleware/Rewrite/src/Internal/PatternSegments/IISServerVariableSegment.cs]: nit: let's make this internal.

File: src/Middleware/Rewrite/src/Internal/IISUrlRewrite/ServerVariables.cs
Comment (jkotalik) [inline src/Middleware/Rewrite/src/Internal/IISUrlRewrite/ServerVariables.cs]: This is a small a breaking change if one of these variables return a different value than the fallback did and there is no way to turn it off. We may need to add an option or something to not use IISServerVariables.

File: src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.csproj
Comment (mderriey) [reply src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.csproj]: This your call. Would we relocate the interface to a common project and then have both `.Server.IIS` and `Rewrite` depend on it?

I'll leave it this way for now, but feel free to drop another comment here if you decide you don't want the current dependency.

Comment (Tratcher) [reply src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.csproj]: How old is the feature? Moving it is likely breaking.

The dependencies aren't really a problem anymore for anything in the shared framework.

File: src/Middleware/Rewrite/src/Internal/IISUrlRewrite/ServerVariables.cs
Comment (mderriey) [reply src/Middleware/Rewrite/src/Internal/IISUrlRewrite/ServerVariables.cs]: True.

Going up the layers, I think we can add an optional boolean parameter to the [`AddIISUrlRewrite`](https://github.com/aspnet/AspNetCore/blob/b0659f660b6e8151a10da1c6c16310ec0101c380/src/Middleware/Rewrite/src/IISUrlRewriteOptionsExtensions.cs) extension methods and let it flow down to `ServerVariables.FindServerVariable`.

Comment (mderriey) [reply src/Middleware/Rewrite/src/Internal/IISUrlRewrite/ServerVariables.cs]: Rook a shot at introducing a way to opt out in [`2e90d8f3`](https://github.com/aspnet/AspNetCore/commit/2e90d8f3c9b71834196f854ca3f6d64ad48ef7fd)

File: src/Middleware/Rewrite/src/IISUrlRewriteOptionsExtensions.cs
Comment (jkotalik) [inline src/Middleware/Rewrite/src/IISUrlRewriteOptionsExtensions.cs]: I'd invert the bool here and have this on by default. I'm fine with this being a breaking change, however it would be bad if there wasn't a work around.

File: src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.csproj
Comment (anurse) [reply src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.csproj]: I agree, I don't think the dependency issue is blocking. I just wanted to raise it. The shared framework does make this point mostly moot. There's no scenario where Rewrite is available and Server.IIS isn't.

Ok, let's go ahead then.

File: src/Middleware/Rewrite/src/IISUrlRewriteOptionsExtensions.cs
Comment (mderriey) [reply src/Middleware/Rewrite/src/IISUrlRewriteOptionsExtensions.cs]: Yeah, that makes sense. Did that in [`9fedfd53`](https://github.com/aspnet/AspNetCore/pull/10022/commits/9fedfd5330a8261cc3e2d970cdf9f32f7c3b11f0).

I was wondering whether to change the default of the parameter or rename it to `disableIISServerVariables`.

I ended up keeping the original name, change the default value and update the associated comment, but let me know if you think renaming the parameter would make more sense to you.

**Comment by davidfowl** [comment]
I think we should move this feature to the features assembly in Http/*

File: src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.csproj
Comment (davidfowl) [reply src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.csproj]: Lets move it and type forward.

Comment (davidfowl) [reply src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.csproj]: It doesn't need to block this PR tho

**Comment by mderriey** [comment]
Hey @davidfowl thanks for jumping in.

What feature are you referring to exactly?

File: src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.csproj
Comment (Tratcher) [reply src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.csproj]: @davidfowl we're going to have trouble expanding the features model if we insist that every feature interface must go into a central dll. Also consider that feature interfaces can bring with them additional types specific to their scenario (e.g. IResponseCookiesFeature, IResponseCookies, CookieOptions).  Cases like this where the feature is component specific warrant locating the feature interface closer to that component. What about adding new packages like IIS.Abstractions? That gives you some separation but still give us a growth model. 

A more extreme example would be if a component that wasn't in the shared runtime wanted to expose a feature interface. It wouldn't make sense to add a feature to Http.Features for a component we wouldn't even bundle with the runtime.

Comment (davidfowl) [reply src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.csproj]: I understand the concern and I share it. I don't think all features should go into there (for e.g. I'm not suggesting we move all of the kestrel internal features) but server variables are generic enough to imagine supporting them outside of IIS and dependency free enough that I think it's reasonable to move it to the features assembly. 

We don't have ANIs anymore...

**Comment by Tratcher** [comment]
@mderriy he's talking about moving IServerVariablesFeature so we can avoid the IIS server dependency. We'd also need to move GetIISServerVariable, or not use it in this code path.

**Comment by anurse** [comment]
That would be a breaking change, but I think I'm OK with it. It seems reasonable. Server Variables aren't actually an IIS-specific concept anyway. They date back to [CGI](https://en.wikipedia.org/wiki/Common_Gateway_Interface) so it seems reasonable that other servers may want to provide them.

If we do this, I'd consider renaming the extension method to `GetServerVariable` instead of `GetIISServerVariable`.

**Comment by davidfowl** [comment]
> That would be a breaking change, but I think I'm OK with it. It seems reasonable. Server Variables aren't actually an IIS-specific concept anyway. They date back to CGI so it seems reasonable that other servers may want to provide them.

Exactly!

> If we do this, I'd consider renaming the extension method to GetServerVariable instead of GetIISServerVariable.

Right! I think that's what we want to do for 3.0. Lets not hold up this PR as a result tho. File an issue and refactor this since most of these changes are related to the rewrite middleware.

**Comment by mderriey** [comment]
Thanks for the explanation/context.

I'd be happy to try and implement the final solution to expand my knowledge of the codebase.

Could you please tag me in the resulting issue when it's created?

**Comment by anurse** [comment]
Filing a bug now :)

**Comment by anurse** [comment]
https://github.com/aspnet/AspNetCore/issues/10167

Feel free to move discussion on the subject of moving/renaming the feature to that issue.

**Comment by mderriey** [comment]
Great, thanks. I'll wait until this PR is merged to start the move.

Thanks again.

**Comment by anurse** [comment]
Code Check failed but I can't get AzDO to load. Usually this means you've got reference assemblies that are out of date. See https://github.com/aspnet/AspNetCore/blob/master/docs/ReferenceAssemblies.md for instructions on updating.

**Comment by anurse** [comment]
Got it to load, yep, just need to update the refasms.

```
error: Generated code is not up to date in src/Middleware/Rewrite/ref/Microsoft.AspNetCore.Rewrite.netcoreapp3.0.cs. You might need to regenerate the reference assemblies or project list (see docs/ReferenceAssemblies.md and docs/ReferenceResolution.md)
```

File: src/Middleware/Rewrite/src/IISUrlRewriteOptionsExtensions.cs
Comment (jkotalik) [inline src/Middleware/Rewrite/src/IISUrlRewriteOptionsExtensions.cs]:
```suggestion
        public static RewriteOptions AddIISUrlRewrite(this RewriteOptions options, IFileProvider fileProvider, string filePath, bool alwaysUseManagedServerVariables = false)
```

Comment (jkotalik) [reply src/Middleware/Rewrite/src/IISUrlRewriteOptionsExtensions.cs]: Generally we prefer bool defaults to be false.

File: src/Middleware/Rewrite/test/IISUrlRewrite/TestServerVariablesFeature.cs
Comment (jkotalik) [inline src/Middleware/Rewrite/test/IISUrlRewrite/TestServerVariablesFeature.cs]: nit copyright.

**Review by jkotalik — APPROVED** [review]
Nits at this point 😄 

Looks good, we can handle moving the feature to Http in a second PR. 

I also want to do some E2E verification that this works when running inprocess.

File: src/Middleware/Rewrite/src/IISUrlRewriteOptionsExtensions.cs
Comment (mderriey) [reply src/Middleware/Rewrite/src/IISUrlRewriteOptionsExtensions.cs]: No worries, thanks for pointing that out. I pushed extra commits to address that.

File: src/Middleware/Rewrite/test/IISUrlRewrite/TestServerVariablesFeature.cs
Comment (mderriey) [reply src/Middleware/Rewrite/test/IISUrlRewrite/TestServerVariablesFeature.cs]: Fixed 😇

**Comment by jkotalik** [comment]
There are still test failures and I believe you need to re update references. That's the reason we are working on making things internal.

**Comment by mderriey** [comment]
Hey @jkotalik, sorry about that. I forgot to invert the boolean values after we changed the meaning of the parameter. I also regenerated reference assemblies.

Sorry that this one drags out so much, I keep forgetting about the ref assemblies.

Thanks for your time, I'll keep an eye on the outcomes of the builds.

**Comment by jkotalik** [comment]
Thanks @mderriey

**Pull request merged**
